[Headless Bots] Parse Games in parallel

RoiEXLab opened the pull request:
Adresses #2555
Instead of loading all games sequentially, they are now loaded in parallel (for bots, because they always load all games on boot).
This doubles the speed of parsing all maps on my system.

prastle commented:
YAY!

ssoloff commented on src/main/java/games/strategy/engine/framework/headlessGameServer/AvailableGames.java:
```diff
@@ -34,12 +38,11 @@
-  private final TreeMap<String, URI> availableGames = new TreeMap<>();
-  private final Set<String> availableMapFolderOrZipNames = new HashSet<>();
+  private final Map<String, URI> availableGames = Collections.synchronizedMap(new TreeMap<>());
+  private final Set<String> availableMapFolderOrZipNames = Collections.synchronizedSet(new HashSet<>());
```
It seems that once an `AvailableGames` instance is created, it is never again accessed concurrently.  Therefore, it shouldn't be necessary to use thread-safe collections for the entire lifetime of an `AvailableGames` instance.  They only need to be thread safe while `populateAvailableGames()` is running.

My suggestion would be to separate the initialization of these collections from the `AvailableGames` instance by using a factory method:

```diff
diff --git a/src/main/java/games/strategy/engine/framework/headlessGameServer/AvailableGames.java b/src/main/java/games/strategy/engine/framework/headlessGameServer/AvailableGames.java
index 4b53f1552..e6dd6fa42 100644
--- a/src/main/java/games/strategy/engine/framework/headlessGameServer/AvailableGames.java
+++ b/src/main/java/games/strategy/engine/framework/headlessGameServer/AvailableGames.java
@@ -38,11 +38,20 @@ import games.strategy.util.UrlStreams;
 public class AvailableGames {
   private static final boolean delayedParsing = false;
   private static final String ZIP_EXTENSION = ".zip";
-  private final Map<String, URI> availableGames = Collections.synchronizedMap(new TreeMap<>());
-  private final Set<String> availableMapFolderOrZipNames = Collections.synchronizedSet(new HashSet<>());
+  private final Map<String, URI> availableGames;
+  private final Set<String> availableMapFolderOrZipNames;
 
-  AvailableGames() {
-    populateAvailableGames(availableGames, availableMapFolderOrZipNames, Collections.synchronizedSet(new HashSet<>()));
+  private AvailableGames(final Map<String, URI> availableGames, final Set<String> availableMapFolderOrZipNames) {
+    this.availableGames = new TreeMap<>(availableGames);
+    this.availableMapFolderOrZipNames = new HashSet<>(availableMapFolderOrZipNames);
+  }
+
+  static AvailableGames newInstance() {
+    final Map<String, URI> availableGames = Collections.synchronizedMap(new TreeMap<>());
+    final Set<String> availableMapFolderOrZipNames = Collections.synchronizedSet(new HashSet<>());
+    final Set<String> mapNamePropertyList = Collections.synchronizedSet(new HashSet<>());
+    populateAvailableGames(availableGames, availableMapFolderOrZipNames, mapNamePropertyList);
+    return new AvailableGames(availableGames, availableMapFolderOrZipNames);
   }
 
   List<String> getGameNames() {
diff --git a/src/main/java/games/strategy/engine/framework/headlessGameServer/HeadlessGameServer.java b/src/main/java/games/strategy/engine/framework/headlessGameServer/HeadlessGameServer.java
index 7a08d05b9..9cff1f4b8 100644
--- a/src/main/java/games/strategy/engine/framework/headlessGameServer/HeadlessGameServer.java
+++ b/src/main/java/games/strategy/engine/framework/headlessGameServer/HeadlessGameServer.java
@@ -393,7 +393,7 @@ public class HeadlessGameServer {
       logger.info("Running ShutdownHook.");
       shutdown();
     }));
-    m_availableGames = new AvailableGames();
+    m_availableGames = AvailableGames.newInstance();
     m_gameSelectorModel = new GameSelectorModel();
     final String fileName = System.getProperty(GameRunner.TRIPLEA_GAME_PROPERTY, "");
     if (fileName.length() > 0) {
```

You could avoid the defensive copies in the constructor by refactoring a bit more to make the `populateFromDirectory()` and `populateFromFile()` methods to accept a `lock` parameter, but the above is probably the simplest thing that will work.

RoiEXLab replied:
I actually found an even easier solution...
See https://github.com/triplea-game/triplea/pull/2571/commits/d047201f9ee4e6138282944d0b24e12b7b988115

ssoloff replied:
Yes, that was what I was thinking about.  However, it wasn't clear to me that it is 100% thread safe because there doesn't appear to be a memory barrier to ensure the thread executing `AvailableGames#AvailableGames()` sees the effects of the background threads.  The act of copying a synchronized collection would trigger such a memory barrier and guarantee the constructor thread deterministically sees the final state of those collections.

It appears that `Future#get()` (which is invoked for all background threads via `invokeAll()`) has a _happens-before_ guarantee, which implies it effectively sets up a memory barrier between all background threads and the constructor thread.  So, in its current state, this _should_ be thread safe.

My only remaining suggestion in this area would be to encapsulate all the concurrency-related code in `populateAvailableGames()` so it's localized in one place, and a code reader doesn't have to back up 50 lines to see that these collections are thread-safe.  Also, the third parameter is for internal use by the method and can be removed:

```diff
diff --git a/src/main/java/games/strategy/engine/framework/headlessGameServer/AvailableGames.java b/src/main/java/games/strategy/engine/framework/headlessGameServer/AvailableGames.java
index 5ea70b285..ca959f1eb 100644
--- a/src/main/java/games/strategy/engine/framework/headlessGameServer/AvailableGames.java
+++ b/src/main/java/games/strategy/engine/framework/headlessGameServer/AvailableGames.java
@@ -41,10 +41,7 @@ public class AvailableGames {
   private final Set<String> availableMapFolderOrZipNames = new HashSet<>();
 
   AvailableGames() {
-    populateAvailableGames(
-        Collections.synchronizedMap(availableGames),
-        Collections.synchronizedSet(availableMapFolderOrZipNames),
-        Collections.synchronizedSet(new HashSet<>()));
+    populateAvailableGames(availableGames, availableMapFolderOrZipNames);
   }
 
   Set<String> getGameNames() {
@@ -77,8 +74,11 @@ public class AvailableGames {
     return Optional.ofNullable(availableGames.get(gameName)).map(Object::toString).orElse(null);
   }
 
-  private static void populateAvailableGames(final Map<String, URI> availableGames,
-      final Set<String> availableMapFolderOrZipNames, final Set<String> mapNamePropertyList) {
+  private static void populateAvailableGames(final Map<String, URI> outAvailableGames,
+      final Set<String> outAvailableMapFolderOrZipNames) {
+    final Map<String, URI> availableGames = Collections.synchronizedMap(outAvailableGames);
+    final Set<String> availableMapFolderOrZipNames = Collections.synchronizedSet(outAvailableMapFolderOrZipNames);
+    final Set<String> mapNamePropertyList = Collections.synchronizedSet(new HashSet<>());
     System.out.println("Parsing all available games (this could take a while). ");
     final List<File> files = allMapFiles();
     if (files.size() > 0) {
```

There's probably better choices for the parameter names.

ssoloff commented on src/main/java/games/strategy/engine/framework/headlessGameServer/AvailableGames.java:
```diff
@@ -75,11 +78,28 @@ String getGameFilePath(final String gameName) {
+        }
+      }
+      service.shutdown();
+      for (Future<?> future : tasks) {
```
Any reason why you can't just call `service.awaitTermination()` here?  It doesn't appear that you are using any of the task results and are simply using them to join the current thread on the worker threads, which `awaitTermination()` should do, as well.

Alternatively, since this is a one-off executor service anyway, you could just submit all the tasks at once using `invokeAll()`, which doesn't return until all tasks are complete, and avoid `shutdown()`, `awaitTermination()`, `get()`, etc.

RoiEXLab replied:
awaitTermination requires a timeout parameter, which would then need to be set to an arbitrary high value with no real reasoning.
Using invokeAll is probably the best idea.

ssoloff replied:
Agreed that `invokeAll()` is probably the better choice here.

FWIW, the Javadocs for the [`java.util.concurrent`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/package-summary.html) package states:

>  All methods that accept timeout parameters treat values less than or equal to zero to mean not to wait at all. **To wait "forever", you can use a value of `Long.MAX_VALUE`.**

It's always nice to get an official blessing from Oracle on what would normally seem like a hack. :smile:

RoiEXLab replied:
Agreed 👍

RoiEXLab commented:
@ssoloff Note that https://github.com/triplea-game/triplea/pull/2571/commits/85fb42c11266099dbc606af0b7b082a96699dd14 contains some unrelated changes that simply avoid creating multiple copies of collections

ssoloff commented on src/main/java/games/strategy/engine/framework/headlessGameServer/AvailableGames.java:
```diff
@@ -75,11 +80,30 @@ String getGameFilePath(final String gameName) {
+    final List<File> files = allMapFiles();
+    if (files.size() > 0) {
+      final ExecutorService service = Executors.newWorkStealingPool(files.size());
+      final List<Callable<Object>> tasks = new ArrayList<>(files.size());
```
It's a shame there's no `invokeAll()` overload for a collection of `Runnable`s.  Changing the generic type parameter to `Void` might better convey the intention that there is no return value from these tasks.

RoiEXLab replied:
Good suggestion, my first attempt was using <?> as type, but error-prone complained so I needed to change it to object

ssoloff commented on src/main/java/games/strategy/engine/framework/headlessGameServer/AvailableGames.java:
```diff
@@ -75,11 +80,30 @@ String getGameFilePath(final String gameName) {
-        populateFromZip(map, availableGames, availableMapFolderOrZipNames, mapNamePropertyList);
+    final List<File> files = allMapFiles();
+    if (files.size() > 0) {
+      final ExecutorService service = Executors.newWorkStealingPool(files.size());
```
This is creating one thread per installed map.  I guess that could be fairly large for a bot (~135?).  Just to be safe, should we limit the maximum number of threads that can be created during this operation?

Map parsing is probably a predominantly CPU-bound operation, but there will be some portion that is IO-bound due to reading the fairly-large zip files from disk.  Without going through the trouble of calculating a [blocking coefficient](http://baddotrobot.com/blog/2013/06/01/optimum-number-of-threads/) for this task, I'm wondering if `# of cores + fixed #` would be sufficient.  `fixed #` would be a guessing game just to keep the CPU busy while the next map is being read.  Without measuring, I'm thinking `fixed #` is probably between 1-2.

RoiEXLab replied:
What's your suggestion?

ssoloff replied:
You could try both `Runtime.getRuntime().availableProcessors() + 1` and `Runtime.getRuntime().availableProcessors() + 2` with a large number of maps and see which one yields better average performance.

RoiEXLab commented:
Result of my Benchmarks (time in nanoseconds, average values, the Map Files are stored on an SSD Samsung 850 EVO 250GB, run count 10 times, all maps, AMD FX-8350 8 Cores):
Current settings (Work Stealing Pool= amount of maps): 2.84200278178E10
Work Stealing Pool Cores +1: 2.71759910686E10
Fixed Thread Pool Cores +1: 2.69100581148E10
Work Stealing Pool Cores +2: 2.65485107534E10
Fixed Thread Pool Cores +2: 2.59100483789E10
Fixed Thread Pool Cores +3: 2.60192406995E10
Fixed Thread Pool Cores * 2: 2.63774260206E10

Extra Test with  Fixed Thread Pool Cores + 2 and the map.isDirectory/isFile checks inside the multithreaded task: 2.66753192583E10
So the option is clear now!

Feel free to do your own benchmarking though, and adjust the value if neccessary!
